### ISSUE TITLE
False positive when dict with literal keys unpacked as function args

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#8671]
**Describe the bug**

If a dictionary annotated with literal keys is passed to a function with matching string arguments, if there are other defaulted arguments of a different type there is an error that the value type in the dictionary can't be used for that argument.

**Code or Screenshots**

```python
from typing import Literal, TypedDict

def func(*, foo: str, bar: str, baz: int | None = None):
    ...

func_args: dict[Literal["foo", "bar", "qux"], str] = {
    "foo": "foo",
    "bar": "bar",
    "qux": "bar",
}

class FuncArgs(TypedDict):
    foo: str
    bar: str
    qux: str

# incorrect error
# Argument of type "str" cannot be assigned to parameter "baz" of type "int | None" in function "func"
func(**func_args)
# No error, perhapse should be
func(**func_args, baz=None)
# correct error
# Unable to match unpacked TypedDict argument to parameters
func(**FuncArgs(**func_args))
```

It's not a big issue as explicitly passing the defaulted argument fixes the problem. A `TypedDict` also works, but `TypedDict` derives from `Mapping[str, object]` so can't be used anywhere that a narrower `Mapping` is expected.

**VS Code extension or command-line**
pylance/pyright cli/playground

**Comment by erictraut** [comment]
I think pyright's behavior is correct here, so I don't consider this a bug.

The "incorrect error" case that you mentioned is a legitimate type violation. If an error were not generated here, it would enable `func` to be called with a value for `baz` that violates the type declaration for that parameter.

The "No error, perhaps should be" case is OK because the type signature of `func` cannot be violated in this case. It is possible in this case for a "duplicate parameter" runtime error, but that's different from a type violation.

**Comment by erictraut** [comment]
I just realized that I misread your `func_args` definition, so my analysis was wrong. Reopening for further investigation.

**Comment by alwaysmpe** [comment]
The literal annotation of string keys means that the `func_args` dict cannot contain a value for `baz` without also triggering a type violation

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.376.
